go/mysql/master_status.go

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	replicationdatapb "vitess.io/vitess/go/vt/proto/replicationdata"
+)
+
+// MasterStatus holds replication information from SHOW MASTER STATUS.
+type MasterStatus struct {
+	// Position represents the master's GTID based position.
+	Position Position
+	// FilePosition represents the master's file based position.
+	FilePosition Position
```

Comment:
This is confusing to me. In https://github.com/vitessio/vitess/pull/6365/files#diff-177f395ad90cb369abe43e7b694045f6R140-R141, and in https://github.com/vitessio/vitess/pull/6365/files#diff-8bdd98f98735260494e4979edf01e69cR238, it appears like both `Position` and `FilePosition` can, and _do_ contain either GTID or file/pos data. In such case, would it make sense for `MasterStatus` to only have a single field called `Position`, and for the type `Position` to have a flag/enum stating whether this is file/pos or GTID?

Reply:
In this case both fields will always be filled in. I agree that it's a bit confusing because `Position` can be either, but I also think refactoring `Position` is out of scope for this particular PR.

Reply:
The main goal of this PR is to make DemoteMaster always return the "legacy", file-based position as well as the "native", flavor-specific (usually GTID-based) position.

In some cases of the new ERS design, we will need to fall back to comparing file-based positions even for flavors that normally use GTIDs. We may not know we need to fall back until after all the positions have been fetched, so for simplicity I proposed that we just always return both positions ("native" and "legacy") so the information is always there if we need it.

Reply:
> refactoring Position is out of scope for this particular PR.

Fair.